errors/context.go

```diff
@@ -0,0 +1,193 @@
+package errors
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/mailgun/holster/v5/callstack"
+	"github.com/sirupsen/logrus"
+)
+
+// HasContext Implement this interface to pass along unstructured context to the logger
+type HasContext interface {
+	Context() map[string]interface{}
+}
+
+// HasFormat True if the interface has the format method (from fmt package)
+type HasFormat interface {
+	Format(st fmt.State, verb rune)
+}
+
+// WithContext Creates errors that conform to the `HasContext` interface
+type WithContext map[string]interface{}
+
+// Wrapf returns an error annotating err with a stack trace
+// at the point Wrapf is call, and the format specifier.
+// If err is nil, Wrapf returns nil.
+func (c WithContext) Wrapf(err error, format string, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   err,
+		msg:     fmt.Sprintf(format, args...),
+	}
+}
+
+// Wrap returns an error annotating err with a stack trace
+// at the point Wrap is called, and the supplied message.
+// If err is nil, Wrap returns nil.
+func (c WithContext) Wrap(err error, msg string) error {
+	if err == nil {
+		return nil
+	}
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   err,
+		msg:     msg,
+	}
+}
+
+func (c WithContext) Error(msg string) error {
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   errors.New(msg),
+		msg:     "",
+	}
+}
+
+func (c WithContext) Errorf(format string, args ...interface{}) error {
+	return &withContext{
+		stack:   callstack.New(1),
+		context: c,
+		cause:   fmt.Errorf(format, args...),
+		msg:     "",
+	}
+}
+
+// Implements the `error` `causer` and `Contexter` interfaces
+type withContext struct {
+	context WithContext
+	msg     string
+	cause   error
+	stack   *callstack.CallStack
+}
+
+func (c *withContext) Unwrap() error {
+	return c.cause
+}
+
+func (c *withContext) Is(target error) bool {
```

Comment:
You do not need an explicit `Is` if `Unwrap` is defined, also it does not do anything special that `errors.Is` does not do itself.

Reply:
@horkhe 
`Unwrap() error`
and
`Is(error) bool`
are different interfaces - https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/errors/wrap.go

Reply:
From my experience, `Is(target)` allows `errors.Is()` to correctly identify the error in a stack of wrapped errors. I'll try removing it and see if the tests pass.